### ISSUE TITLE
Paste as snippet command to be able to use variables (like $1)

### ISSUE THREAD

**Issue by Rokuem** [microsoft/vscode#81853]
Managing snippets in vscode is bothersome since you need to edit them on a JSON file (specially because of spacing, strings, suggestions), I tried using some snippets managers (extensions and softwares), but they lack the integration with the snippets variables that we can use in vscode, like $1 to define where the cursor should go.

I think the fastest and easiest solution would be:

A command to paste as a snippet, so vscode would get the clipboard content, interpret it as a snippet and paste the result. This would be the easiest solution to allow external snippets to use the vscode features. We wouldn't have the suggestions for those external snippets, but at least we could use macros, share one use snippets, and extensions could adapt more easily. The conflict here would be with the clipboard variable, which could be ignored, deleted or have a interaction like: enter to add snippet again or tab to skip it (could be useful for repetition, like imports of modules with same variable name as the module. In this case we would have a repetition variable and the clipboard one would fallback to it when the paste as snippet is used).

Other suggestions would be:

- Have an UI with a code editor to manage snippets in vscode. (would solve the initial problem and could also have nice features, like testing the snippet, listing the variables we can use and etc... But our snippets would be limited to vscode. Some people may prefer using a cloud service or something of the sort).
- Allow the loading of snippets .json files based on a file pattern list in the settings. (so others programs could integrate with it more easily at least)

**Comment by jrieken** [comment]
> A command to paste as a snippet, so vscode would get the clipboard content, interpret it as a snippet and paste the result.

You can (and should) do that as extension. Tho, I don't understand where the snippet would come from in the first place?

**Comment by Rokuem** [comment]
Not sure if you are saying that I can create this command with a extension or that I should use snippets extensions instead. For a extension for it to be possible I would need to be able to use the vscode snippet interpreter (for cursor positions and etc.) and have access to the clipboard content of the user, haven't created a extension yet, but can take a look in case those things are possible (are they?). But otherwise: 
---------------------------
tl;dr: We can have extensions to improve the experience, some of them do, but haven't found a good one to manage the already created snippets (in a good way, instead of JSON) and that supports variables at the same time. but the problem will still be there so features like voice commands, mouse gestures and macros won't be able to integrate well without this "paste as snippet command".

So, answering your question, the snippets would come from those macro programs and also from snippets managers software and cloud services. 

-------

@jrieken For example:
https://github.com/lauthieb/code-notes

While I can use a external program like this one to manage snippets, I wont be able to create a simple snippet with variables like:

```
ìmport $0 from "$0";
```
This is something easy to create on vscode, just using this small snippet for the sake of demonstration, more complex ones would be things like:
- Creating a database model (where the model name repeats in some parts)
- Creating a vue component (where there is the main element class, which is also the name on the default export and also the main selector on the stylesheet, specially if you use something like BEM naming for the classes)

Creating it and placing the variables is not a big problem on vscode, but editing them as you change the way you do things is very painful to do in JSON. It's usually easier to create it again.

This can be improved with an extension, by either modifying the snippets file or using a custom interpreter and prompting for the variables values. For some reason haven't seem a extension that does it well yet, might be missing something, would love to be wrong. There are some that create snippets from selection and others that you can get snippets from a cloud provider (where they prompt you for the name) and also the snippets bundle extensions, but none of them are good to use. (does not solve the edit in JSON problem and it's hard to know the snippets well)

Some other things that could also be implemented by the extension, but usually would be easier to have in another software are:
- Cloud integration for a more controlled type of access to the snippets (ex: project based, user based, role based, etc...).
- Collaboration
- Tags
- Editor for the already created snippets (instead of using JSON)

But while those solutions improve the experience on vscode, it is still hard to use macro programs like:
- StrokePlus (mouse gestures, macros, text expansions, all with custom scripting).
- Nuance Dragon (voice commands with scripting).

As a use case example, while I can use my hands just fine, I wanted to try using things like Nuance Dragon and saying "create import", then it would paste the import snippet and I could just type / or say the name of the module and press tab to go to the line bellow (on this step I could make the program insert the snippet again with scripting too for example)

**Comment by jrieken** [comment]
Not sure how marcos, voice commands, or mouse gestures relate to this. But here are some APIs: for inserting snippets (and putting the editor in snippet mode): https://github.com/microsoft/vscode/blob/229df29bc9bb6d7e2a731c1dbe879933fd8745cb/src/vs/vscode.d.ts#L1133. No need for implementing a custom snippet interpreter. Then you can also return snippets from [completion item providers](https://github.com/microsoft/vscode/blob/229df29bc9bb6d7e2a731c1dbe879933fd8745cb/src/vs/vscode.d.ts#L8356)

**Comment by Rokuem** [comment]
@jrieken nice! Ty for the help :), will try to dig a little into it when I have some some, but seems easy to do with the `insertSnippet` function.

To clarify about the macros:
![image](https://user-images.githubusercontent.com/26756006/66116880-dfe3e900-e5a9-11e9-9dfc-1e660aaea11c.png)

This, for example, is a very simple command to trigger a hotkey on the specified gesture (in this case it's associated to a integrated terminal creation).

The most basic idea would be to set the clipboard content then execute the vscode command to paste as snippet :).

not very familiar with the scripting on nuance dragon yet, but for this one (strokePlus) it would be something like this:

```
clip.SetText('ìmport $0 from "$0"');
sp.SendModifiedVKeys([vk.LCONTROL,vk.LSHIFT], [vk.VK_V]); // the shortcut for the command on vscode.
```

There are many advantages of being able to do things like this, some of them are:
- Can use it globally (useful for writing examples, documentations, etc... Provided there is a paste fallback when you are not on vscode).
- Some things are easier to do with voice / gestures / text expansions (which can be triggered with space too)
- Extra automation
- Accessibility (since people that code with voice would be able to use it)
